src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/Threads.java

```diff
@@ -144,39 +144,36 @@ private static synchronized void initialize(TypeDataBase db) {
         }
 
         virtualConstructor = new VirtualConstructor(db);
-        // Add mappings for all known thread types
+
+        /*
+         * Add mappings for JavaThread types
+         */
+
         virtualConstructor.addMapping("JavaThread", JavaThread.class);
+
         if (!VM.getVM().isCore()) {
             virtualConstructor.addMapping("CompilerThread", CompilerThread.class);
         }
-        virtualConstructor.addMapping("JvmtiAgentThread", JvmtiAgentThread.class);
-        virtualConstructor.addMapping("ServiceThread", ServiceThread.class);
-        virtualConstructor.addMapping("MonitorDeflationThread", MonitorDeflationThread.class);
-        virtualConstructor.addMapping("NotificationThread", NotificationThread.class);
-        virtualConstructor.addMapping("StringDedupThread", StringDedupThread.class);
-        virtualConstructor.addMapping("AttachListenerThread", AttachListenerThread.class);
 
-        /* Only add DeoptimizeObjectsALotThread if it is actually present in the type database. */
+        // These are all the visible JavaThread subclasses that execute java code.
+        virtualConstructor.addMapping("JvmtiAgentThread", JavaThread.class);
+        virtualConstructor.addMapping("NotificationThread", JavaThread.class);
+        virtualConstructor.addMapping("AttachListenerThread", JavaThread.class);
```

Comment:
It still seems a shame that we have to enumerate all the thread sybtypes.

Reply:
In order the create a mirror instance, SA needs to map the hotspot type to the SA type that implements the mirror. We have the same issue with CodeBlob subtypes.